### ISSUE TITLE
Methods exposed to Windows Runtime must return only Windows Runtime types.

### ISSUE THREAD

**Issue by JanWerder** [tomkuijsten/restup#45]
I'm getting a 

```
'AlarmClock.ParameterController.GetWithSimpleParameters(System.Int32, System.String)' returns 'Devkoes.Restup.WebServer.Models.Schemas.GetResponse', which is not a valid Windows Runtime type.  Methods exposed to Windows Runtime must return only Windows Runtime types.
````
when following the REST tutorial in the wiki. I'm also using a headless app and getting the same results when trying the example provided in the repo.

**Comment by Jark** [comment]
Hi,

Can you let us know:

- what device you're trying it on? Is it Windows 10, Raspberry PI, Beagleboard ?
- what version of the W10 sdk are you using? (Right-click a project and look at the Target Version)
- do you get these errors on start-up or when you call the rest method via a browser

Thanks,

Jark

**Comment by JanWerder** [comment]
* Raspberry Pi 3
* Target Version 10.0 Build 10586
* When I compile the solution

**Comment by Jark** [comment]
I've done a clean clone from the repository and both the HeadlessDemo and HeadedDemo seem to work fine on a Raspberry PI 2.

I suspect the issue seems to be with your setup of UWP in Visual Studio, have you followed the steps on http://ms-iot.github.io/content/en-US/win10/GetStarted/SetUpVisualStudio.htm? Also, you might need the insider sdk if you're using the Raspberry PI 3 (https://dev.windows.com/en-US/downloads/windows-10-developer-preview?OCID=insider).

Are you able to compile / deploy a completely new and empty headless app?

**Comment by JanWerder** [comment]
- I follwed the steps from the link you provided before and I'm able to deploy an empty solution.
- Are you sure about the insider SDK? It seems like a language problem more than a SDK problem

I'll attach a screenshot in case that might help you:
![](http://i.imgur.com/ewLDo85.png)

**Comment by Jark** [comment]
Hello Jan,

Thanks, that did help me with pinpointing what's going on.

I think (don't know for sure since I haven't read up on the uwp compiler model) that all types exported in windows uwp "apps" have to be sealed.

Why the sample app works is because it has got the controllers in a separate library, and this library is referenced by the sample app. Once the sample app is compiled all the library's types are made sealed (if you're not overriding them).

What this means:
- If you want to use restup now you will have to use a separate library project and reference the controllers from that library in the headless app, or override the GetResponse class and make that class sealed
- the response classes in restup will need to be made sealed to prevent this from happening. I probably will have some time to do this this weekend, but pull requests are also always welcome :)

I want to apologise for suggesting that your setup was wrong, and thank you for following this through!

Cheers,

Jark

**Comment by JanWerder** [comment]
Hey Jark,

Thanks for the follow up! I appreciate your support.
Since a seperate library project seems a bit over the top to me, if this is something that can be fixed, I think I'll wait. I would like to help out, but I'm not informed enough to put together a PR for this issue.

Thanks ,
Jan

**Comment by Jark** [comment]
Hello Jan,

After looking into this it turns out the change will be a bit more involved than I'd like.

I think I need to:

- refactor all the responses:
  - move all the types that are defined within them to the outer scope
  - remove all logic that uses generics from them (T IRestResponse.Visit<P, T>(IRestResponseVisitor<P, T> visitor, P param)
  - make them sealed
  - remove the rest verb attribute and move that information into the responses
- make the async controller stuff work / check that it works with IAsyncOperation since that's not allowed in uwp components
- modify DemoControllers for it to be a Windows Runtime Component

There's also a chance that I will have to move the responses + enums + interfaces they use to a separate project, but I'd like to avoid that if possible.

So it will be a bit longer before I'll have a pull request ready. I haven't got an eta yet, except for that it will be at some point this week.

@tomkuijsten do you agree with me doing this? Do we want to support the scenario where you can directly add restup to your Windows IOT background app?

**Comment by JanWerder** [comment]
Hey Jark,

Glad to hear! 
I hope @tomkuijsten agrees, but I would be suprised to hear otherwise since his [blog indicates](http://www.devkoes.nl/windows-iot-background-app/) that he is working on an iot project as well.

**Comment by Jark** [comment]
Hey Jan,

Yeah, he is using it for his own project. But I think he defines his controllers in a "Class library" and then references that in a windows runtime component. That scenario works as expected.

This change is to make it possible to define the controllers in a windows runtime component directly.

**Comment by JanWerder** [comment]
Ok, thanks for clearing that up, but that would require the inclusion of the restup sourcecode into the developers solution, wouldn't it? That may work, but is definitly not the way a library should work. Ideally it would just require the developer to check out the lastet version from nuget.

But that's why the changes described above are neccessary, right? :wink:

**Comment by tomkuijsten** [comment]
Hmm. I do like the idea that you could use it directly in a project. You are correct about me implementing the controllers in another project, that's probably why I didn't bump into this problem.

@Jark could you explain to me what public contracts will be changed?

**Comment by Jark** [comment]
@JanWerder No, you wouldn't need the restup source, you'd just need to reference restup in both your class library and main project, and define the controllers in the class library.

@tomkuijsten 
These "contracts":
- GetResponse
- PostResponse
- PutResponse
- DeleteResponse
[Now have their internal types (enums) made public (had to rename them in the process) and are turned into a class (since structs can't do inheritance in windows runtime components)](https://github.com/Jark/restup/commit/9a784fc344e85bf679b07030f607ba9748122a7f).

[I've removed the Visit method from IRestResponse since generics are not supported in interfaces.](https://github.com/Jark/restup/commit/9c7675b4743f4b87121fa74e0060f1164d42d78f).

I envisage only the following classes / interfaces need to be in the windows runtime component:
- IRestResponse.cs
- IContentRestResponse
- BadRequestResponse
- DeleteResponse
- DeleteResponseStatus
- GetResponse
- GetResponseStatus
- MethodNotAllowedResponse
- PostResponse
- PostResponseStatus
- PutResponse
- PutResponseStatus
- StatusOnlyResponse

I still need to:
- make them sealed to make them valid WinRT classes
- put them into the namespace: `Devkoes.Restup.WebServer.Rest.Response` for the responses and the interfaces (and the name of the .winmd file), this does break compatibility but it will be easy to fix.

Let me know what you think.

**Comment by JanWerder** [comment]
Is this closed with the merged PR commit 4b3d86c ?
If so, would it be possible to push that version to nuget?

**Comment by tomkuijsten** [comment]
There are a couple of things I like to do before we push another version to nuget, eg remove some obsolete files. Until then you could build the bits yourself. If you need any instructions on how to do that, let me know.

**Comment by Jark** [comment]
Closed due to #46
